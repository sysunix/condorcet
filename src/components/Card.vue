<template>
  <div class="px-6 rounded overflow-hidden shadow-md">
    <div class="pt-4">
      <div class="font-bold text-xl mb-3">{{ question }}</div>
      <p class="text-gray-700 text-lg mb-2">
        {{ $tc("poll.participants", users.length, { count: users.length }) }}
      </p>
      <p class="text-gray-700 text-base">
        {{ description }}
      </p>
    </div>
    <div class="flex justify-between flex-wrap  py-3">
      <router-link
        v-if="users.includes(userId)"
        :data-testid="
          isActive ? `vote-${questionSlug}` : `results-${questionSlug}`
        "
        class="bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded w-full my-1"
        tag="button"
        :to="{
          name: isActive ? 'poll_show' : 'poll_result',
          params: { id: id }
        }"
        >{{ isActive ? $t("global.vote") : $t("global.results") }}</router-link
      >
      <button
        v-if="users.includes(userId)"
        class="bg-yellow-500 hover:bg-yellow-700 text-white font-bold py-2 px-4 rounded w-full my-1"
        @click.prevent="$emit('onShare', id)"
      >
        {{ $t("global.share") }}
      </button>
      <button
        v-if="owner === userId"
        :data-testid="
          isActive ? `close-${questionSlug}` : `reOpen-${questionSlug}`
        "
        class="bg-green-500 hover:bg-green-700 text-white font-bold py-2 px-4 rounded w-full my-1"
        @click.prevent="$emit('onToggleStatus', id)"
      >
        {{ isActive ? $t("global.close") : $t("global.reOpen") }}
      </button>

      <button
        v-if="users.includes(userId) === false"
        class="bg-purple-500 hover:bg-purple-700 text-white font-bold py-2 px-4 rounded w-full my-1"
        @click.prevent="$emit('onJoin', id)"
      >
        {{ $t("global.join") }}
      </button>

      <button
        v-if="users.includes(userId) === true && owner !== userId"
        class="bg-gray-500 hover:bg-gray-700 text-white font-bold py-2 px-4 rounded w-full my-1"
        @click.prevent="$emit('onLeave', id)"
      >
        {{ $t("global.leave") }}
      </button>

      <button
        v-if="owner === userId"
        data-testid="delete"
        class="bg-red-500 hover:bg-red-700 text-white font-bold py-2 px-4 rounded w-full my-1"
        @click.prevent="$emit('onDelete', id)"
      >
        {{ $t("global.delete") }}
      </button>
    </div>
  </div>
</template>

<script lang="ts">
import { Vue, Component, Prop } from "vue-property-decorator";
import slugify from "slugify";

@Component
export default class Card extends Vue {
  @Prop({ type: String, required: true }) readonly id!: string;
  @Prop({ type: String, required: true }) readonly userId!: string;
  @Prop({ type: String, required: true }) readonly description!: string;
  @Prop({ type: String, required: true }) readonly question!: string;
  @Prop({ type: String, required: true }) readonly owner!: string;
  @Prop({ type: Boolean, required: true }) readonly isActive!: boolean;
  @Prop({ type: Array, required: true }) readonly users!: Array<string>;

  get questionSlug() {
    return slugify(this.question);
  }
}
</script>

<style></style>
