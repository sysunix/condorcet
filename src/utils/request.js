import firebase, { db } from "../firebase";

export const createPoll = async poll => {
  const pollWithTimestap = {
    ...poll,
    timestamp: firebase.firestore.FieldValue.serverTimestamp()
  };

  return db.collection("polls").add(pollWithTimestap);
};

export const createVote = async (vote, pollId, userId) => {
  const voteWithTimestap = {
    ...vote,
    timestamp: firebase.firestore.FieldValue.serverTimestamp()
  };

  return db
    .collection("polls")
    .doc(pollId)
    .collection("votes")
    .doc(userId)
    .set(voteWithTimestap);
};
