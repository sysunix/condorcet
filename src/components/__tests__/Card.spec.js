import { shallowMount } from "@vue/test-utils";
import Card from "../Card.vue";
import { getTranslation } from "../../utils/test";
import { MOCK_POLL, MOCK_USER_ID, MOCK_POLL_ID } from "../../utils/test";

describe("Card.vue", () => {
  it("should render when there is no notifications", () => {
    const wrapper = shallowMount(Card, {
      propsData: { ...MOCK_POLL, userId: MOCK_USER_ID, id: MOCK_POLL_ID },
      stubs: ["router-link"],
      mocks: {
        $tc: jest.fn(),
        $t: getTranslation
      }
    });
    expect(wrapper).toMatchSnapshot();
  });
});
