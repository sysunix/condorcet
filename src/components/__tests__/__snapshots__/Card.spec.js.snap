// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card.vue should render when there is no notifications 1`] = `
<div class="rounded overflow-hidden shadow-md" answers="[object Object],[object Object],[object Object]" ispublic="true" timestamp="[object Object]" token="MOCK_TOKEN">
  <div class="px-6 py-4">
    <div class="font-bold text-xl mb-3">Qui est le meilleur ?</div>
    <p class="text-gray-700 text-lg mb-2">

    </p>
    <p class="text-gray-700 text-base">Qui repellendus est aliquam vel qui.</p>
  </div>
  <div class="flex justify-between flex-wrap px-6 py-4">
    <router-link-stub data-testid="vote-Qui-est-le-meilleur" tag="button" to="[object Object]" class="bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded w-full my-1">Voter</router-link-stub> <button class="bg-yellow-500 hover:bg-yellow-700 text-white font-bold py-2 px-4 rounded w-full my-1">
      Partager
    </button> <button data-testid="close-Qui-est-le-meilleur" class="bg-green-500 hover:bg-green-700 text-white font-bold py-2 px-4 rounded w-full my-1">
      Fermer
    </button>
    <!---->
    <!----> <button data-testid="delete" class="bg-red-500 hover:bg-red-700 text-white font-bold py-2 px-4 rounded w-full my-1">
      Supprimer
    </button>
  </div>
</div>
`;
